cloud/cluster/memory/subscription.go

```diff
@@ -0,0 +1,56 @@
+package memory
+
+import (
+	"github.com/scootdev/scoot/cloud/cluster"
+)
+
+// subscriber promptly reads its input and maintains a queue so that the sender
+// doesn't have to worry about a send blocking.
+type subscriber struct {
+	inCh    chan []cluster.NodeUpdate
+	outCh   chan []cluster.NodeUpdate
+	closeCh chan interface{}
+	queue   []cluster.NodeUpdate
+}
+
+func makeSubscription(initial []cluster.Node, closeCh chan interface{}, inCh chan []cluster.NodeUpdate) cluster.Subscription {
+	s := &subscriber{
+		inCh:    inCh,
+		outCh:   make(chan []cluster.NodeUpdate),
+		closeCh: closeCh,
+		queue:   nil,
+	}
+	go s.loop()
```

Comment:
If we pretend this isn't golang for a second, would you raise your eyebrows
if a java constructor created a thread? Maybe it's something I have to get
used to in this language, but I think it's clearly different than how we
expect other languages to be written.

On Mon, Jul 18, 2016 at 10:09 AM, Dan Bentley <notifications@github.com>
wrote:

> In cloud/cluster/memory/subscription.go
> <https://github.com/scootdev/scoot/pull/60#discussion_r71187710>:
>
> > +// doesn't have to worry about a send blocking.
> > +type subscriber struct {
> > +	inCh    chan []cluster.NodeUpdate
> > +	outCh   chan []cluster.NodeUpdate
> > +	closeCh chan interface{}
> > +	queue   []cluster.NodeUpdate
> > +}
> > +
> > +func makeSubscription(initial []cluster.Node, closeCh chan interface{}, inCh chan []cluster.NodeUpdate) cluster.Subscription {
> > +	s := &subscriber{
> > +		inCh:    inCh,
> > +		outCh:   make(chan []cluster.NodeUpdate),
> > +		closeCh: closeCh,
> > +		queue:   nil,
> > +	}
> > +	go s.loop()
>
> No, you shouldn't have to read the impl!
>
> But as a user, why would you care that this creates a goroutine? You
> wouldn't care that it allocates memory.
>
> —
> You are receiving this because you commented.
> Reply to this email directly, view it on GitHub
> <https://github.com/scootdev/scoot/pull/60/files/e861c728ad8f30e0c6caea09c8d5803bfee06bfb#r71187710>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/ABc5CVaD675UqMYHR1ifQm2RFzaWlkLkks5qW7NhgaJpZM4JNyBt>
> .
>